asterius/src/Asterius/Ld.hs

```diff
@@ -45,9 +46,13 @@ data LinkTask
 loadTheWorld :: LinkTask -> IO AsteriusCachedModule
 loadTheWorld LinkTask {..} = do
   ncu <- newNameCacheUpdater
-  lib <- mconcat <$> for linkLibs (loadAr ncu)
-  objs <- rights <$> for linkObjs (tryGetFile ncu)
+  lib <- parallelFor 1 linkLibs (loadAr ncu) -- TODO: Parameterize
```

Comment:
`loadAr` is already parallelized, and here we use `parallelFor` again, which creates nested parallelism. Not to say this won't run, but the overall efficiency will be reduced.

We need to either:

* Refactor `loadAr`, so that it only returns the `ArchiveEntry`s, and after we collect all the `ArchiveEntry`s, parallelize their deserialization (along with `objs`)
* Improve our parallelization framework so that it supports nested parallelism; each job can spawn new jobs to be evaluated somewhere else.

Reply:
I went with the first approach:

> Refactor `loadAr`, so that it only returns the `ArchiveEntry`s, and after we collect all the `ArchiveEntry`s, parallelize their deserialization (along with `objs`)

but I don't completely understand what you mean with the second bullet:

> Improve our parallelization framework so that it supports nested parallelism; each job can spawn new jobs to be evaluated somewhere else.

Nested parallelism is already allowed, right? I understand that nested calls would pin threads to the same capabilities (hence slower), but it'd still work. What is the alternative solution you mean here?

Reply:
Ah, the second approach elaborated is:

* We have a `P` monad which you can lift `IO` actions into. It carries the context (thread pool, task queue) needed for all parallelism
* There's a `fork :: P a -> P (Future a)` method which spawns a `P a` action for parallel execution. The `P a` action itself can `fork` new jobs, thus nested parallelism here.
* There's a `join :: Future a -> P a` method which blocks until the result is available.

Going through all the trouble of creating the `P` monad abstraction has an advantage: we can enforce the number of worker threads to match CPU core count all the time. Directly calling `parallelFor` inside `parallelFor` would be much less efficient.